### PR TITLE
Scalar point implementation

### PR DESCRIPTION
**Pull request by KiriosK**

### CODE REVIEW

File: bn254/src/fq.rs [hunk 1]
```diff
@@ -184,7 +184,7 @@ prime_field_operation!(Fq, MODULUS, GENERATOR, INV, R, R2, R3);
 
 impl From<Fr> for Fq {
     fn from(val: Fr) -> Fq {
-        Self(val.0)
+        Self(to_mont_form(val.montgomery_reduce(), R2, MODULUS, INV))
```
Comment (ashWhiteHat) [inline bn254/src/fq.rs:187]: I think we don't perform reduce either of them

File: bn254/src/fr.rs [hunk 2]
```diff
@@ -293,7 +293,7 @@ fft_field_operation!(
 
 impl From<Fq> for Fr {
     fn from(val: Fq) -> Fr {
-        Self(to_mont_form(val.0, R2, MODULUS, INV))
+        Self(to_mont_form(val.montgomery_reduce(), R2, MODULUS, INV))
```
Comment (ashWhiteHat) [inline bn254/src/fr.rs:296]: I think we don't perform reduce either of them

File: r1cs/src/gadget/binary.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    #[test]
+    fn binary_assignment_instance() {
+        let mut cs: R1cs<GrumpkinDriver> = R1cs::default();
+
+        let _ = BinaryAssignment::instance(&mut cs, 0);
+        let _ = BinaryAssignment::instance(&mut cs, 1);
+        let _ = BinaryAssignment::witness(&mut cs, 0);
+        let _ = BinaryAssignment::witness(&mut cs, 1);
+
+        assert!(cs.is_sat());
```
Comment (ashWhiteHat) [inline r1cs/src/gadget/binary.rs:48]: We don't check anything so we can remove this test

File: r1cs/src/gadget/binary.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    #[test]
+    #[should_panic]
+    fn binary_assignment_instance_non_bit() {
```
Comment (ashWhiteHat) [inline r1cs/src/gadget/binary.rs:53]: We don't check anything so we can remove this test

File: r1cs/src/gadget/binary.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
+use crate::driver::CircuitDriver;
+use crate::{R1cs, Wire};
+use std::marker::PhantomData;
+
+#[derive(Clone)]
+pub struct BinaryAssignment<C: CircuitDriver>(Wire, PhantomData<C>);
```
Comment (ashWhiteHat) [inline r1cs/src/gadget/binary.rs:6]: We can replace CircuitDriver with PrimeField

File: r1cs/src/gadget/field.rs [hunk 6]
```diff
@@ -57,11 +62,62 @@ impl<C: CircuitDriver> FieldAssignment<C> {
         z
     }
 
+    fn range_check(cs: &mut R1cs<C>, value: &Self, num_of_bits: u16) {
```
Comment (ashWhiteHat) [inline r1cs/src/gadget/field.rs:65]: Is the way to check binary decomposition correct?
I think we don't use inner_product but minus one check as following article says.
https://hackmd.io/@7dpNYqjKQGeYC7wMlPxHtQ/B1w_9nq2Y

**Review by ashWhiteHat — CHANGES_REQUESTED** [review]

File: bn254/src/fq.rs (hunk 1 above)
Comment (KiriosK) [reply bn254/src/fq.rs:187]: I'll check if it's possible to do without reduction.

File: r1cs/src/gadget/binary.rs (hunk 3 above)
Comment (KiriosK) [reply r1cs/src/gadget/binary.rs:48]: These tests just automatically check if the 0, 1 value is passed to the assignment and if the wrong one is declined.
If it's not necessary, I can remove it.

File: r1cs/src/gadget/binary.rs (hunk 5 above)
Comment (KiriosK) [reply r1cs/src/gadget/binary.rs:6]: Will fix it.

Comment (KiriosK) [reply r1cs/src/gadget/binary.rs:6]: We use R1cs<C>

File: r1cs/src/gadget/binary.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            panic!("Bit value should be passed, got {bit}");
+        }
+        let wire = cs.public_wire();
+        cs.x.push(C::Scalar::from(bit as u64));
+
+        Self(wire, PhantomData::default())
+    }
+
+    pub fn witness(cs: &mut R1cs<C>, bit: u8) -> Self {
+        if bit != 0 && bit != 1 {
```
Comment (ashWhiteHat) [inline r1cs/src/gadget/binary.rs:20]: we also check whether this bit is binary so we don't have to constrain with panic

File: r1cs/src/gadget/field.rs (hunk 6 above)
Comment (KiriosK) [reply r1cs/src/gadget/field.rs]: Fixed

File: r1cs/src/gadget/binary.rs (hunk 3 above)
Comment (KiriosK) [reply r1cs/src/gadget/binary.rs]: Removed

File: r1cs/src/gadget/binary.rs (hunk 7 above)
Comment (KiriosK) [reply r1cs/src/gadget/binary.rs]: Removed

File: r1cs/src/gadget/binary.rs (hunk 4 above)
Comment (KiriosK) [reply r1cs/src/gadget/binary.rs]: Removed

**Pull request merged**